src/LevelZeroIOGroup.cpp

```diff
@@ -0,0 +1,906 @@
+            if (m_signal_pushed[ii] == signal) {
+                result = ii;
+                is_found = true;
+            }
+        }
+        if (!is_found) {
+            // If not pushed, add to pushed signals and configure for batch reads
+            result = m_signal_pushed.size();
+            m_signal_pushed.push_back(signal);
+            signal->setup_batch();
+
+            if (m_special_signal_set.find(signal_name) != m_special_signal_set.end()) {
+                push_signal(signal_name + "_TIMESTAMP", domain_type, domain_idx);
+            }
+        }
+
+        return result;
+    }
+
+    // Mark the given control to be written by write_batch()
+    int LevelZeroIOGroup::push_control(const std::string &control_name,
+                                       int domain_type, int domain_idx)
+    {
+        if (!is_valid_control(control_name)) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": control_name " + control_name +
+                            " not valid for LevelZeroIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_type != control_domain_type(control_name)) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": " + control_name + ": domain_type must be " +
+                            std::to_string(control_domain_type(control_name)),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 ||
+            domain_idx >= m_platform_topo.num_domain(domain_type)) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": domain_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        int result = -1;
+        bool is_found = false;
+        std::shared_ptr<control_s> control = m_control_available.at(
+                                             control_name).controls.at(domain_idx);
+
+        // Check if control was already pushed
+        for (size_t ii = 0; !is_found && ii < m_control_pushed.size(); ++ii) {
+            // same location means this control or its alias was already pushed
+            if (m_control_pushed[ii] == control) {
+                result = ii;
+                is_found = true;
+            }
+        }
+        if (!is_found) {
+            // If not pushed, add to pushed control
+            result = m_control_pushed.size();
+            m_control_pushed.push_back(control);
+        }
+
+        return result;
+    }
+
+    // Parse and update saved values for signals
+    void LevelZeroIOGroup::read_batch(void)
+    {
+        m_is_batch_read = true;
+        for (size_t ii = 0; ii < m_signal_pushed.size(); ++ii) {
+            std::dynamic_pointer_cast<LevelZeroSignal>(m_signal_pushed[ii])->
```

Comment:
Be nice if we could avoid this cast here.  Not sure why m_signal_pushed needs to be a Signal pointer and not a LevelZeroSignal pointer.